### PR TITLE
Optional X-Consumer-Groups ACL Header Improvement

### PR DESCRIPTION
**Pull request by jeremyjpj0916**
### Summary

As discussed here : https://github.com/Kong/kong/issues/3609

One of my biggest concerns with Kong's ACL functionality as it stands today is including the arbitrarily growing in size list of a consumers entire ACL groups as a header. In our specific implementation of Kong every route gets is own ACL group as the routes existing uuid, so we can then give consumers access to specific routes by adding that uuid to their ACL, I suspect this will be a popular pattern going forward for many users who implement Kong. 

I propose a boolean that will act as a switch and enable/disable sending the consumer groups as a header, as I suspect for plenty of use cases that the arbitrary group info in the header is not necessary and as it grows in size could one day break an otherwise working proxy by increasing the header size past what a API Provider will accept (since many API providers have a 8kb header size limit by default like Apache Tomcat). I would like a predictable header size going to the API provider that is static when it comes to size generally, and by enabling this functionality on the ACL plugin we can achieve that with Kong.

### Full changelog

1. Modify schema.lua to include new boolean property 'hide_groups_header' defaulted to false as to not break existing functionality

2. Modify handler.lua to account for 'hide_groups_header' boolean setting.

### Issues resolved

Fixes concern for growing Kong header size as ACL groups list grow. Also improves bytes across the wire to exclude potentially extraneous info of a consumers entire ACL group list.

---

Let me know if I am missing anything, not super familiar with the PR process in general but would like to see this functionality in one of Kong's upcoming releases as I have a few Kong consumer accounts at 50+ ACL groups and that extra header is already irking me a bit 😄 .

### CODE REVIEW

File: kong/plugins/acl/schema.lua [hunk 1]
```diff
@@ -4,7 +4,8 @@ return {
   no_consumer = true,
   fields = {
     whitelist = { type = "array" },
-    blacklist = { type = "array" }
+    blacklist = { type = "array" },
+    set_groups_header = { type = "boolean", default = true }
```
Comment (thibaultcha) [inline kong/plugins/acl/schema.lua]: The default for this value should probably respect the already existing behaviour, that is to _include_ the groups header. Thus, `default = false`. Now, to semantically align with the authentication plugins, which provide a `hide_credentials` option (whose default is `false`), this option should probably be updated to `hide_groups_header = { type = "boolean", default = false }`. Thoughts?

Comment (jeremyjpj0916) [reply kong/plugins/acl/schema.lua]: And then just do this right?
```
  if not conf.hide_groups_header then
    set_header(constants.HEADERS.CONSUMER_GROUPS, to_be_blocked)
  end
```
This works fine for me 👍 , whatever Kong prefers honestly. Glad to see yall are cool with adding this functionality.

Comment (thibaultcha) [reply kong/plugins/acl/schema.lua]: @jeremyjpj0916 Yes, exactly.

**Comment by jeremyjpj0916** [comment]
Added related documentation PR here: https://github.com/Kong/docs.konghq.com/pull/821

**Comment by jeremyjpj0916** [comment]
@thibaultcha We attempted to rebase to master... but doing it without a new PR from the UI. This is not my strong suit, I am not a smart man 😆 ...

**Comment by jeremyjpj0916** [comment]
I think I did a proper testcase here: https://github.com/Kong/kong/pull/3697/commits/e004a02149b3215ee1f8f1548bfc580ab88d878d , I include basic test for if true then make sure that header value equals nil, I also include a if false make sure the header is still accounted for.

**Comment by jeremyjpj0916** [comment]
Based on my reading here it seems its true you cannot change the source branch of a PR: https://stackoverflow.com/questions/42381557/changing-source-branch-for-pr-on-github , why don't I just close this one unimplemented and make a newer one since that seems easiest?

**Comment by thibaultcha** [comment]
@jeremyjpj0916 You can change the base branch for your PR if you click on the "edit" button on top of it. You can then chose "master", and rebase/cherry-pick/etc... your commit(s) on top of the master branch, and force push the branch. Or reopen a PR is fine too if that's too much trouble.

**Comment by jeremyjpj0916** [comment]
Sorry to disappoint man 😅 , but I do understand the latter half method, sorry for the extra overhead closing and opening a new one adds but its easiest on me at least.

image

All I see to do so in UI is to point to different target. Lets give this one more go on a diff PR. I assure it will be perfecto!

**Comment by thibaultcha** [comment]
Yeah that is the right button
